phq/kafka/consumer.py

```diff
@@ -47,6 +47,14 @@ def _group_messages(messages):
     return batch_ref
 
 
+def _latest_distinct_messages(messages):
+    latest_distinct_msgs = OrderedDict()
+    for msg in messages[::-1]:
+        if msg.id not in latest_distinct_msgs:
+            latest_distinct_msgs[msg.id] = msg
+    return list(latest_distinct_msgs.values())[::-1]
```

Comment:
To follow our new convention with FP, maybe more something like:
```
def _latest_distinct_messages(messages):
    unique_msg = OrderedDict(
        (message.id, message)
        for message in messages
    )
    return list(unique_msg.values())
```
This will always have only the latest value, without having to reorder the list twice.

Reply:
but the order might not be the same as your version in this case, we could still do something like:
```
def _latest_distinct_messages(messages):
    unique_msg = OrderedDict(
        (message.id, message)
        for message in messages[::-1]
    )
    return list(unique_msg.values())[::-1]
```
Just the complexity of it seems a bit to much (we have an OrderDict + to reorder of the list)

Reply:
But In my version, the order will not be right, as it will keep the order of the last inserted element, here another thought:
```python
# used from python documentation
class LastUpdatedOrderedDict(OrderedDict):
    """Store items in the order the keys were last added"""

    def __setitem__(self, key, value):
        super().__setitem__(key, value)
        self.move_to_end(key)

def _latest_distinct_messages(messages):
    unique_msg = LastUpdatedOrderedDict(
        (message.id, message)
        for message in messages
    )
    return list(unique_msg.values())
```

Reply:
There are many ways to solve this type of thing and you can spend hours on StackOverflow to find the fanciest and most performant solution that may not be as readable. 

From my point of view, I am not a fan of creating LastUpdatedOrderedDict class that calls move_to_end() func. That code also runs at least 4x slower than my version after I ran multiple unit tests (100x) with different message sizes. So, I am keeping my implementation in this case.